sklearn/cluster/_hdbscan/_tree.pyx

```diff
@@ -442,46 +413,41 @@ cdef cnp.ndarray[cnp.intp_t, ndim=1, mode='c'] do_labelling(
     cdef:
         cnp.intp_t root_cluster
         cnp.ndarray[cnp.intp_t, ndim=1, mode='c'] result
-        cnp.intp_t[:] parent_array, child_array
-        cnp.float64_t[:] lambda_array
+        cnp.ndarray[cnp.intp_t, ndim=1] parent_array, child_array
+        cnp.ndarray[cnp.float64_t, ndim=1] lambda_array
         TreeUnionFind union_find
-        cnp.intp_t n, parent, child, cluster
+        cnp.intp_t n, parent, child, cluster, label
 
-    child_array = hierarchy['child']
-    parent_array = hierarchy['parent']
-    lambda_array = hierarchy['value']
+    child_array = condensed_tree['child']
+    parent_array = condensed_tree['parent']
+    lambda_array = condensed_tree['value']
 
     root_cluster = np.min(parent_array)
     result = np.empty(root_cluster, dtype=np.intp)
     union_find = TreeUnionFind(np.max(parent_array) + 1)
 
-    for n in range(hierarchy.shape[0]):
+    for n in range(condensed_tree.shape[0]):
         child = child_array[n]
         parent = parent_array[n]
         if child not in clusters:
             union_find.union(parent, child)
 
     for n in range(root_cluster):
         cluster = union_find.find(n)
-        if cluster < root_cluster:
-            result[n] = NOISE
-        elif cluster == root_cluster:
-            if len(clusters) == 1 and allow_single_cluster:
-                if cluster_selection_epsilon != 0.0:
-                    if hierarchy['value'][hierarchy['child'] == n] >= 1 / cluster_selection_epsilon :
-                        result[n] = cluster_label_map[cluster]
-                    else:
-                        result[n] = NOISE
-                elif hierarchy['value'][hierarchy['child'] == n] >= \
-                     hierarchy['value'][hierarchy['parent'] == cluster].max():
-                    result[n] = cluster_label_map[cluster]
-                else:
-                    result[n] = NOISE
-            else:
-                result[n] = NOISE
-        else:
-            result[n] = cluster_label_map[cluster]
-
+        label = NOISE
+        if cluster != root_cluster:
+            label = cluster_label_map[cluster]
+        elif len(clusters) == 1 and allow_single_cluster:
+            if cluster_selection_epsilon != 0.0:
+                if lambda_array[child_array == n] \
+                    >= 1 / cluster_selection_epsilon:
+                    label = cluster_label_map[cluster]
+
+            elif lambda_array[child_array == n] >= \
+                    lambda_array[parent_array == cluster].max():
```

Comment:
Similarly: is it possible to store this condition in a variable and reuse it here?
```suggestion
            elif lambda_array[child_array == n] >= \
                    lambda_array[parent_array == cluster].max():
```